From d00466da2ce8300e2f0b64196af5b582d0e8cb4e Mon Sep 17 00:00:00 2001
From: Takuro Ashie <ashie@clear-code.com>
Date: Mon, 6 Oct 2025 21:35:28 +0900
Subject: [PATCH 17/21] Enable to use gfx::NativePixmapDmaBuf if available on
 Wayland

By this change, OzoneImageBacking become available even on platforms
that don't support GBM (`--in-process-gpu` will be needed on such
platforms). It's needed for hardware video decode on Wayland.

Upstream-Status: Pending

Signed-off-by: Takuro Ashie <ashie@clear-code.com>
---
 ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc | 6 +++++-
 ui/ozone/platform/wayland/gpu/wayland_surface_factory.h  | 2 +-
 ui/ozone/platform/wayland/ozone_platform_wayland.cc      | 4 ++--
 3 files changed, 8 insertions(+), 4 deletions(-)

diff --git a/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc b/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc
index c409f5c7a8..ecbdad5caa 100644
--- a/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc
+++ b/ui/ozone/platform/wayland/gpu/wayland_surface_factory.cc
@@ -312,11 +312,15 @@ WaylandSurfaceFactory::CreateNativePixmapFromHandle(
   return nullptr;
 }
 
-bool WaylandSurfaceFactory::SupportsNativePixmaps() const {
+bool WaylandSurfaceFactory::SupportsNativePixmaps(bool in_process_gpu) const {
   bool supports_native_pixmaps = false;
 #if defined(WAYLAND_GBM)
   supports_native_pixmaps = buffer_manager_->GetGbmDevice() != nullptr;
 #endif
+  if (!supports_native_pixmaps && in_process_gpu) {
+    supports_native_pixmaps =
+      gl::GLSurfaceEGL::GetGLDisplayEGL()->ext->b_EGL_EXT_image_dma_buf_import;
+  }
   // Native pixmaps are not supported with swiftshader.
   if (gl::IsSoftwareGLImplementation(gl::GetGLImplementationParts())) {
     supports_native_pixmaps = false;
diff --git a/ui/ozone/platform/wayland/gpu/wayland_surface_factory.h b/ui/ozone/platform/wayland/gpu/wayland_surface_factory.h
index 93c954d3f6..fe5ed0fbe0 100644
--- a/ui/ozone/platform/wayland/gpu/wayland_surface_factory.h
+++ b/ui/ozone/platform/wayland/gpu/wayland_surface_factory.h
@@ -58,7 +58,7 @@ class WaylandSurfaceFactory : public SurfaceFactoryOzone {
   void SetDrmModifiersFilter(
       std::unique_ptr<DrmModifiersFilter> filter) override;
 
-  bool SupportsNativePixmaps() const;
+  bool SupportsNativePixmaps(bool in_process_gpu) const;
 
   std::vector<gfx::BufferFormat> GetSupportedFormatsForTexturing()
       const override;
diff --git a/ui/ozone/platform/wayland/ozone_platform_wayland.cc b/ui/ozone/platform/wayland/ozone_platform_wayland.cc
index 458598dd89..39b1ae575f 100644
--- a/ui/ozone/platform/wayland/ozone_platform_wayland.cc
+++ b/ui/ozone/platform/wayland/ozone_platform_wayland.cc
@@ -425,7 +425,7 @@ class OzonePlatformWayland : public OzonePlatform,
       if (surface_factory_) {
         DCHECK(has_initialized_gpu());
         properties.supports_native_pixmaps =
-            surface_factory_->SupportsNativePixmaps();
+            surface_factory_->SupportsNativePixmaps(single_process());
       }
 
       properties.supports_global_application_menus =
@@ -442,7 +442,7 @@ class OzonePlatformWayland : public OzonePlatform,
           buffer_manager_->supports_overlays() &&
           buffer_manager_->supports_viewporter();
       properties.supports_native_pixmaps =
-          surface_factory_->SupportsNativePixmaps();
+          surface_factory_->SupportsNativePixmaps(single_process());
     }
     return properties;
   }
-- 
2.43.0

